include/cuda/std/stream

```diff
@@ -0,0 +1,156 @@
+   * outlive the stream identified by the `cudaStream_t` handle.
+   *
+   */
+  explicit constexpr stream_view(cudaStream_t stream) : __stream{stream} {}
+
+  /// Disallow construction from an `int`, e.g., `0`.
+  stream_view(int) = delete;
+
+  /// Disallow construction from `nullptr`
+  stream_view(std::nullptr_t) = delete;
+
+  /// Returns the wrapped `cudaStream_t` handle
+  constexpr cudaStream_t value() const noexcept { return __stream; }
+
+  /// Explicit conversion to `cudaStream_t` returns the wrapped `cudaStream_t`
+  /// handle
+  explicit constexpr operator cudaStream_t const noexcept { return value(); }
+
+  /**
+   * @brief Synchronizes the wrapped stream.
+   *
+   * @throws TODO: 
+   *
+   */
+  void synchronize() const { cudaStreamSynchronize(value()); }
+
+private:
+  cudaStream_t __stream{0}; ///< Handle of the viewed stream
+};
+
+/**
+ * @brief Compares two `stream_view`s for equality
+ *
+ * @param lhs The first `stream_view` to compare
+ * @param rhs The second `stream_view` to compare
+ * @return true if equal, false if unequal
+ */
+inline constexpr bool operator==(stream_view lhs, stream_view rhs) {
+  return lhs.value() == rhs.value();
+}
+
+/**
+ * @brief Compares two `stream_view`s for inequality
+ *
+ * @param lhs The first `stream_view` to compare
+ * @param rhs The second `stream_view` to compare
+ * Return true if unequal, false if equal
+ */
+inline constexpr bool operator!=(stream_view lhs, stream_view rhs) {
+  return not(lhs == rhs);
+}
+
+/**
+ * @brief An owning wrapper of a `cudaStream_t`.
+ *
+ * `unique_stream` is movable, but not copyable.
+ *
+ */
+class unique_stream {
+public:
+  /**
+   * @brief Construct a `unique_stream` wrapping a new CUDA stream.
+   *
+   * TODO: Flags, priority
+   */
+  unique_stream() : __stream{new cudaStream_t} {
+    cudaStreamCreate(__stream.get());
+  }
+
+  ~unique_stream() = default;
+  unique_stream(unique_stream const &) = default;
+  unique_stream(unique_stream &&) = default;
+  unique_stream &operator=(unique_stream const &) = default;
+  unique_stream &operator=(unique_stream &&) = default;
+
+  /**
+   * @brief Returns a `stream_view` of the owned stream.
+   *
+   * @note: It is the callers responsibility to ensure the resulting
+   * `stream_view` does not outlive the `unique_stream`.
+   *
+   * @throws //TODO:
+   *
+   * @return stream_view
+   */
+  operator stream_view() const {
+    // Check if the owned stream has been moved
+    if (__stream) {
+      return stream_view{*__stream};
+    }
+    throw std::runtime_error; // TODO: Throw a proper exception here
+  }
+
+  /**
+   * @brief Synchronizes the owned stream.
+   * 
+   * @throws //TODO:
+   *
+   */
+  void synchronize() const { stream_view{*this}.synchronize(); }
+
+private:
+  struct destroy_stream {
+    void operator()(cudaStream_t *__s) { cudaStreamDestroy(*__s); }
+  };
+  std::unique_ptr<cudaStream_t, destroy_stream> __stream;
```

Comment:
We should use unique_resource (but it's not part of the standard yet) or, knowing that cudaStream_t is some kind of pointer (and it is) we could have:
```suggestion
  struct destroy_stream {
    void operator()(cudaStream_t __s) { cudaStreamDestroy(__s); }
  };
  std::unique_ptr<std::remove_pointer<cudaStream_t>, destroy_stream> __stream;
```

Reply:
We should not punch through the user-facing typedef.

Reply:
I agree it's uglyish. Another option is to forego the use of smart pointer and manage the handle manually.
See:
https://github.com/NVIDIA/DALI/blob/master/include/dali/core/unique_handle.h
https://github.com/NVIDIA/DALI/blob/master/include/dali/core/cuda_stream.h

Reply:
On further consideration, there isn't really a need for this PR to contain an owning `cudaStream_t` type. All that's necessary is the `stream_view` non-owning type for the stream-ordered interface. How about we defer the owning type to future work?

Reply:
On further consideration, there isn't really a need for this PR to contain an owning `cudaStream_t` type. All that's necessary is the `stream_view` non-owning type for the stream-ordered interface. How about we defer the owning type to future work?

Reply:
That's true.

Reply:
> We should not punch through the user-facing typedef.

It looks like that's what Thrust does :). But I agree, I don't love it either. 

https://github.com/NVIDIA/thrust/blob/5df7084448aca0ba345740349f1248835d003deb/thrust/system/cuda/detail/future.inl#L183